cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/inference/DFGFunctionSummaries.kt

```diff
@@ -0,0 +1,197 @@
+        val methodName = functionDecl.name
+        // The language and the method name have to match. If a signature is specified, it also has
+        // to match to the one of the FunctionDeclaration, null indicates that we accept everything.
+        val matchingEntries =
+            functionToDFGEntryMap.keys.filter {
+                it.language == languageName &&
+                    methodName.lastPartsMatch(it.methodName) &&
+                    (it.signature == null ||
+                        functionDecl.hasSignature(
+                            it.signature.map { signatureType -> language.objectType(signatureType) }
+                        ))
+            }
+        return if (matchingEntries.size == 1) {
+            // Only one entry => We take this one.
+            functionToDFGEntryMap[matchingEntries.single()]
+        } else if (matchingEntries.filter { it.signature != null }.size == 1) {
+            // Only one entry with a matching signature => We take this one.
+            functionToDFGEntryMap[matchingEntries.single { it.signature != null }]
+        } else if (matchingEntries.isNotEmpty()) {
+            /* There are multiple matching entries. We use the following routine:
+             * First, we filter for existing signatures.
+             * Second, we filter for the most precise class.
+             * If there are still multiple options, we take the longest signature and hope it's the most precise one.
+             */
+            val typeEntryList =
+                matchingEntries
+                    .filter { it.signature != null }
+                    .map {
+                        Pair(
+                            language.parseName(it.methodName).parent?.let { it1 ->
+                                language?.objectType(it1)
+                            },
+                            it
+                        )
+                    }
+            val mostPreciseClassEntries = mutableListOf<FunctionDeclarationEntry>()
+            var mostPreciseType = typeEntryList.first().first
+            var superTypes = getAllSupertypes(mostPreciseType)
+            for (typeEntry in typeEntryList) {
+                if (typeEntry.first == mostPreciseType) {
+                    mostPreciseClassEntries.add(typeEntry.second)
+                } else if (typeEntry.first in superTypes) {
+                    mostPreciseClassEntries.clear()
+                    mostPreciseClassEntries.add(typeEntry.second)
+                    mostPreciseType = typeEntry.first
+                    superTypes = getAllSupertypes(typeEntry.first)
+                }
+            }
+            if (mostPreciseClassEntries.size > 1) {
+                mostPreciseClassEntries.sortByDescending { it.signature?.size ?: 0 }
+            }
+            functionToDFGEntryMap[matchingEntries.first()]
+        } else {
+            null
+        }
+    }
+
+    private fun getAllSupertypes(type: Type?): Set<Type> {
```

Comment:
I am pretty sure, that we already have a function like this somewhere.

Reply:
Indeed, there was an `ancestor` property/method somewhere. It does not the exact same thing but it's good enough.